File: Assets/_scripts/fish/behaviour/FishWanderingBehaviour.js
class FishWanderingBehaviour extends FishArbitratedBehaviour
{
    var wanderOffset : float = 60;
    var wanderRadiusXZ : float = 10;
    var wanderRadiusY : float = 30;
    var wanderRate : float = 5;
    
    private var invisiblePole : GameObject;
    private var invisibleCarrot : GameObject;
    private var seeking : FishSeekingBehaviour;
    
    function Start(){
        CreatePoleWithCarrot();
        seeking = gameObject.AddComponent(FishSeekingBehaviour);
        seeking.target = invisibleCarrot;        
    }
    
    // function FixedUpdate(){
    //     GetSteering().ApplyTo(gameObject);
    // }

    function GetSteering () : SteeringOutput{
        Profiler.StartProfile(PT.Wandering);
          
        if(!invisiblePole)
            return SteeringOutput.empty;
        UpdatePole();
        ShuffleCarrot();
        
        Profiler.EndProfile(PT.Wandering); 
        
        return seeking.GetSteering();
    }
    
    function SelfDestroy(){
        if(seeking)
            seeking.SelfDestroy();
            
        Destroy(invisiblePole);
        super.SelfDestroy();        
    }
    
    function OnDrawGizmosSelected(){    
        if(!invisiblePole)
            return;
            
        Gizmos.color = Color.white;
        Gizmos.DrawWireCube(invisiblePole.transform.position, invisiblePole.transform.lossyScale * 2);
        Gizmos.color = Color.red;
        Gizmos.DrawSphere(invisibleCarrot.transform.position, Mathf.Min(wanderRadiusXZ / 5, 0.2));        
    }
    
////////////////////////////////////////////////////////////////////////
    function UpdatePole(){
        if(!invisiblePole)
            return;
            
        var fishDirection = transform.forward; 
        fishDirection.y = 0;
        fishDirection = fishDirection.normalized;
        invisiblePole.transform.position = transform.position + fishDirection * wanderOffset;
        
        var selfScale = Vector3(wanderRadiusXZ, wanderRadiusY, wanderRadiusXZ);
        var parentScale = transform.lossyScale;
        var invertedParentScale = Vector3();
        invertedParentScale.x = 1/parentScale.x; invertedParentScale.y = 1/parentScale.y; invertedParentScale.z = 1/parentScale.z;
        invisiblePole.transform.localScale = Vector3.Scale(invertedParentScale, selfScale);
    }

    private function CreatePoleWithCarrot(){
        invisiblePole = GameObject("Pole with carrot");
        invisiblePole.transform.parent = transform;
        UpdatePole();
        
        invisibleCarrot = GameObject("Invisible carrot");        
        invisibleCarrot.transform.parent = invisiblePole.transform;
        invisibleCarrot.transform.localPosition = Vector3.forward;        
    }

    private function ShuffleCarrot(){
        var oldPos = invisibleCarrot.transform.localPosition.normalized;
        var randomRotation = Quaternion.Euler(RandomAngle(), RandomAngle(), RandomAngle());
        var newPos = randomRotation * oldPos;
        invisibleCarrot.transform.localPosition = newPos;
    }
    
    private function RandomAngle() : float{
        return Utils.RandomBinomial() * wanderRate;
    } 
}

File: Assets/_scripts/fish/behaviour/FishWanderingBehaviour.cs
using UnityEngine;
using System.Collections;

public class FishWanderingBehaviour : FishArbitratedBehaviour {
    
    public GenericSeekingBehaviour seeking;
    
    public float wanderOffset = 10f;
    public float wanderRadiusXZ = 1f;
    public float wanderRadiusY = 3f;
    public float wanderRate = 30f;
    public float carrotUpdatePeriod = 0.1f;
    private float lastCarrotUpdate = 0.0f;
 
    private Transform _transform;
    private GameObject invisiblePole;
    private Transform invisiblePoleTransform;
    private GameObject invisibleCarrot;
    private Transform invisibleCarrotTransform;
    
    
    void Awake(){
        children = new FishBehaviour[1]{seeking};        
    }
    
    void Start(){        
        _transform = transform;
        lastCarrotUpdate = Time.time;
		CreatePoleWithCarrot();        
        seeking.target = invisibleCarrot;
    }

    public override SteeringOutput GetSteering (){
        Profiler.StartProfile(PT.Wandering);

        SteeringOutput ret;

        if(!invisiblePole)
            ret = SteeringOutput.empty;
        else{
            UpdatePolePosition();
            TryShuffleCarrot();

            ret = seeking.GetSteering();            
        }
        
        Profiler.EndProfile(PT.Wandering);             

        return ret;
    }

    public override void SelfDestroy(){
        Destroy(invisiblePole);
        base.SelfDestroy();        
    }

	protected override void PrivateDrawGizmosSelected(){	      
        if(!invisiblePole)
            return;

        Gizmos.color = Color.white;
        DrawUnitCubeIn(invisiblePoleTransform);
    }
    
    void DrawUnitCubeIn(Transform t){
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,-1,-1)), t.TransformPoint(new Vector3(-1,-1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,-1,1)), t.TransformPoint(new Vector3(1,-1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,-1,1)), t.TransformPoint(new Vector3(1,-1,-1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,-1,-1)), t.TransformPoint(new Vector3(-1,-1,-1)));
        
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,1,-1)), t.TransformPoint(new Vector3(-1,1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,1,1)), t.TransformPoint(new Vector3(1,1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,1,1)), t.TransformPoint(new Vector3(1,1,-1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,1,-1)), t.TransformPoint(new Vector3(-1,1,-1)));
        
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,-1,-1)), t.TransformPoint(new Vector3(-1,1,-1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(-1,-1,1)), t.TransformPoint(new Vector3(-1,1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,-1,1)), t.TransformPoint(new Vector3(1,1,1)));
        Gizmos.DrawLine(t.TransformPoint(new Vector3(1,-1,-1)), t.TransformPoint(new Vector3(1,1,-1)));        
    }

////////////////////////////////////////////////////////////////////////
    void UpdatePolePosition(){
        if(!invisiblePole)
            return;            

        Vector3 fishDirection = _transform.forward; 
        fishDirection.y = 0;
        fishDirection = fishDirection.normalized;
        invisiblePoleTransform.position = _transform.position + fishDirection * wanderOffset; 
        Quaternion rotation = _transform.rotation;        
        invisiblePoleTransform.rotation = Quaternion.Euler(0, rotation.eulerAngles.y, 0);
    }
    
    void UpdatePoleScale(){
        if(!invisiblePole)
            return;            
        
        Vector3 selfScale = new Vector3(wanderRadiusXZ, wanderRadiusY, wanderRadiusXZ);
        Vector3 parentScale = _transform.lossyScale;
        Vector3 invertedParentScale = Vector3.zero;
        invertedParentScale.x = 1/parentScale.x; invertedParentScale.y = 1/parentScale.y; invertedParentScale.z = 1/parentScale.z;
        invisiblePoleTransform.localScale = Vector3.Scale(invertedParentScale, selfScale);        
    }

    private void CreatePoleWithCarrot(){
        invisiblePole = new GameObject("Pole with carrot");
        invisiblePoleTransform = invisiblePole.transform;
        invisiblePoleTransform.parent = _transform;
        
        invisibleCarrot = new GameObject("Invisible carrot");        
        invisibleCarrotTransform = invisibleCarrot.transform;
        invisibleCarrotTransform.parent = invisiblePoleTransform;
        invisibleCarrotTransform.localPosition = Random.insideUnitSphere;        
        
        UpdatePoleScale();
        UpdatePolePosition();                
    }
    
    private void TryShuffleCarrot(){
        if(Time.time - lastCarrotUpdate > carrotUpdatePeriod){
            ShuffleCarrot();
            lastCarrotUpdate = Time.time;
        }
    }

    private void ShuffleCarrot(){
        Vector3 oldPos = invisibleCarrotTransform.localPosition.normalized;
        Quaternion randomRotation = Quaternion.Euler(RandomAngle(), RandomAngle(), RandomAngle());
        Vector3 newPos = randomRotation * oldPos;
        invisibleCarrotTransform.localPosition = newPos;
    }

    private float RandomAngle(){
        return Utils.RandomBinomial() * wanderRate;
    } 
}